src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,867 @@
+                                metric_idx++;
+                        }
+                        continue;
+                }
+
+                metrics[metric_idx] =
+                    rd_calloc(1, sizeof(opentelemetry_proto_metrics_v1_Metric));
+                data_points[metric_idx] = rd_calloc(
+                    1, sizeof(opentelemetry_proto_metrics_v1_NumberDataPoint));
+
+                serializeMetricData(
+                    rk, NULL, &info[metrics_to_encode[i]], &metrics[metric_idx],
+                    &data_points[metric_idx],
+                    &datapoint_attributes_key_values[metric_idx],
+                    metricValueCalculator, &metric_names[metric_idx], false,
+                    now_ns);
+                metric_idx++;
+        }
+
+        /* Send empty metrics blob if no metrics are matched */
+        if (total_metrics_count > 0) {
+                metrics_repeated.metrics = metrics;
+                metrics_repeated.count   = total_metrics_count;
+
+                scopeMetrics.metrics.funcs.encode = &encode_metric;
+                scopeMetrics.metrics.arg          = &metrics_repeated;
+
+
+                resource_metrics.scope_metrics.funcs.encode =
+                    &encode_scope_metrics;
+                resource_metrics.scope_metrics.arg = &scopeMetrics;
+
+                metrics_data.resource_metrics.funcs.encode =
+                    &encode_resource_metrics;
+                metrics_data.resource_metrics.arg = &resource_metrics;
+        }
+
+        status = pb_get_encoded_size(
+            &message_size, opentelemetry_proto_metrics_v1_MetricsData_fields,
+            &metrics_data);
+        if (!status) {
+                rd_kafka_dbg(rk, TELEMETRY, "RD_KAFKA_TELEMETRY_METRICS_INFO",
+                             "Failed to get encoded size");
+                free_metrics(metrics, metric_names, data_points,
+                             datapoint_attributes_key_values,
+                             total_metrics_count);
+                free_resource_attributes(resource_attributes_key_values,
+                                         resource_attributes_struct,
+                                         resource_attributes_count);
+                return NULL;
+        }
+
+        buffer = rd_malloc(message_size);
+        if (buffer == NULL) {
+                rd_kafka_dbg(rk, TELEMETRY, "RD_KAFKA_TELEMETRY_METRICS_INFO",
+                             "Failed to allocate memory for buffer");
+                free_metrics(metrics, metric_names, data_points,
+                             datapoint_attributes_key_values,
+                             total_metrics_count);
+                free_resource_attributes(resource_attributes_key_values,
+                                         resource_attributes_struct,
+                                         resource_attributes_count);
+
+                return NULL;
+        }
```

Comment:
This is how we can create a buffer and get a void* to write to it externally. There's an assert in rd_malloc, we don't need to verify that memory allocation succeeded.
```suggestion
        rbuf = rd_buf_new(1, message_size);
        rd_buf_write_ensure(rbuf, message_size, message_size);
        message_size = rd_buf_get_writable(rbuf, &buffer);
```

Reply:
Updated